### PR TITLE
91 grpc file attributes

### PR DESCRIPTION
**Pull request by sajald77**
Issue: [Integrating grpc equivalent of "/v1/file/attributes/" #91](https://github.com/0chain/gosdk/issues/91)

### CODE REVIEW

File: zboxcore/sdk/attributesworker.go [hunk 1]
```diff
@@ -43,63 +43,39 @@ func (ar *AttributesRequest) getObjectTreeFromBlobber(
...
+		PathHash:     remoteFilePathHash,
+		Allocation:   ar.allocationTx,
+		ConnectionId: ar.connectionID,
+		Attributes:   ar.attributes,
+	})
 	if err != nil {
-		Logger.Error(blobber.Baseurl,
-			"Error creating update attributes request", err)
+		Logger.Error("could not update object attributes from blobber -" + blobber.Baseurl + " - " + err.Error())
 		return
```
Comment (kushthedude) [inline zboxcore/sdk/attributesworker.go:64]: Can we wrap & return error instead of nil returns

File: zboxcore/sdk/attributesworker.go [hunk 2]
```diff
@@ -43,63 +43,39 @@ func (ar *AttributesRequest) getObjectTreeFromBlobber(
...
-			var respBody []byte
-			if respBody, err = ioutil.ReadAll(resp.Body); err != nil {
-				Logger.Error(blobber.Baseurl, "Reading response: ", err)
-				return nil
-			}
+	updateBlobberObjectAttributesResult := &blobberhttp.UpdateObjectAttributesResponse{}
+	err = json.Unmarshal(respRaw, updateBlobberObjectAttributesResult)
+	if err != nil {
+		Logger.Error(blobber.Baseurl, "Reading response: ", err)
+		return
```
Comment (kushthedude) [inline zboxcore/sdk/attributesworker.go:72]: Same wrapping & return here

File: zboxcore/sdk/attributesworker.go [hunk 3]
```diff
@@ -43,63 +43,39 @@ func (ar *AttributesRequest) getObjectTreeFromBlobber(
...
-				return nil
-			}
-
-			var respBody []byte
-			if respBody, err = ioutil.ReadAll(resp.Body); err != nil {
-				Logger.Error(blobber.Baseurl, "Reading response: ", err)
-				return nil
-			}
+	updateBlobberObjectAttributesResult := &blobberhttp.UpdateObjectAttributesResponse{}
+	err = json.Unmarshal(respRaw, updateBlobberObjectAttributesResult)
```
Comment (kushthedude) [inline zboxcore/sdk/attributesworker.go:69]: Is there a need for this, we are anyway just updating the objectAttribute & showing if it was successfull or not?

File: core/clients/blobberClient/client_integration_test.go [hunk 4]
```diff
@@ -858,4 +859,84 @@ func TestBlobberClient_IntegrationTest(t *testing.T) {
...
+
+			_, err = UpdateObjectAttributes(BlobberAddr, tc.input)
+			if err != nil {
+				if !tc.expectingError {
+					t.Fatal(err)
+				}
+				continue
+			}
+			if tc.expectingError {
+				t.Fatal("expected error")
```
Comment (kushthedude) [inline core/clients/blobberClient/client_integration_test.go:938]: Can we return error here, will be useful for debugging purpose

**Review by kushthedude — CHANGES_REQUESTED** [review]
Changes looks good just some few niche suggestions, please update your branch with latest grpc & we can go ahead

File: core/clients/blobberClient/client.go [hunk 5]
```diff
@@ -270,3 +272,27 @@ func CalculateHash(url string, req *blobbergrpc.CalculateHashRequest) ([]byte, e
...
+		blobbercommon.ClientHeader:          client.GetClientID(),
+		blobbercommon.ClientKeyHeader:       client.GetClientPublicKey(),
+		blobbercommon.ClientSignatureHeader: clientSignature,
+	}))
+
+	updateObjectAttributesResponse, err := blobberClient.UpdateObjectAttributes(grpcCtx, req)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to UpdateObjectAttributes")
+	}
+	return json.Marshal(convert.UpdateObjectAttributesResponseHandler(updateObjectAttributesResponse))
```
[the remark this answers is not in the document]
Comment (sajald77) [reply core/clients/blobberClient/client.go:297]: I think we usually convert the grpc response to http equivalent response, judging from the other implementations.
 We already had the UpdateObjectAttributesResponseHandler in place. I am just following the pattern already in place.

File: zboxcore/sdk/attributesworker.go (hunk 1 above)
Comment (sajald77) [reply zboxcore/sdk/attributesworker.go:64]: it's does return the error still,but i can wrap them.

File: zboxcore/sdk/attributesworker.go (hunk 3 above)
Comment (sajald77) [reply zboxcore/sdk/attributesworker.go:69]: yes, i think we can remove this altogether.

File: core/clients/blobberClient/client_integration_test.go (hunk 4 above)
Comment (sajald77) [reply core/clients/blobberClient/client_integration_test.go:938]: This is the case when we expecting an error, but do not get one.

**Review by kushthedude — APPROVED** [review]
looks great! few minor improvements but lets take them in the later iterations 👍🏽

**Pull request merged**